lib/r10k/forge/module_release.rb

```diff
@@ -65,34 +79,47 @@ def install(target_dir)
         cleanup
       end
 
-      # Download the module release to {#download_path}
+      # Download the module release to {#download_path} and cache to {#tarball_cache_path}
       #
       # @return [void]
       def download
-        logger.debug1 "Downloading #{@forge_release.slug} from #{PuppetForge::Release.conn.url_prefix} to #{@download_path}"
-        @forge_release.download(download_path)
+        if @tarball_cache_path.exist?
+          logger.debug1 "Using cached copy of #{@forge_release.slug} tarball"
+        else
+          logger.debug1 "Downloading #{@forge_release.slug} from #{PuppetForge::Release.conn.url_prefix} to #{@download_path}"
+          @forge_release.download(download_path)
+          FileUtils::mkdir_p(@tarball_cache_root)
+          FileUtils::mv(@download_path, @tarball_cache_path)
+        end
       end
 
-      # Verify the module release downloaded to {#download_path} against the
-      # module release checksum given by the Puppet Forge
+      # Verify the module release cached in {#tarball_cache_path} against the
+      # module release checksum given by the Puppet Forge. On mismatch, remove
+      # the cached copy.
       #
       # @raise [PuppetForge::V3::Release::ChecksumMismatch] The
-      #   downloaded module release checksum doesn't match the expected Forge
+      #   cached module release checksum doesn't match the expected Forge
       #   module release checksum.
       # @return [void]
       def verify
-        logger.debug1 "Verifying that #{download_path} matches checksum #{@forge_release.file_md5}"
-        @forge_release.verify(download_path)
+        logger.debug1 "Verifying that #{@tarball_cache_path} matches checksum #{@forge_release.file_md5}"
```

Comment:
This should download the md5 and store it next to the tarball, and verify using that.